signer/src/blocklist_client.rs

```diff
@@ -0,0 +1,169 @@
+//! # Blocklist Client Module
+//!
+//! This module provides the `BlocklistChecker` trait and its `BlocklistClient` implementation,
+//! which are used to check addresses against a blocklist service. The module's responsibilities
+//! include querying the blocklist API and interpreting the responses to determine if a given
+//! address is blocklisted, along with its associated risk severity.
+
+use crate::config::SETTINGS;
+use async_trait::async_trait;
+use blocklist_client::common::error::Error;
+use blocklist_client::common::BlocklistStatus;
+use reqwest::Client;
+
+const SCREEN_PATH: &str = "/screen";
+
+#[async_trait]
+pub trait BlocklistChecker {
+    /// Checks if the given address is blocklisted.
+    /// Returns `true` if the address is blocklisted, otherwise `false`.
+    async fn is_blocklisted(&self, address: &str) -> Result<bool, Error>;
+}
+
+#[derive(Clone, Debug)]
+pub struct BlocklistClient {
+    base_url: String,
+    client: Client,
+}
+
+#[async_trait]
+impl BlocklistChecker for BlocklistClient {
+    async fn is_blocklisted(&self, address: &str) -> Result<bool, Error> {
+        let url = Self::address_screening_path(&self.base_url, address);
+        let resp = self
+            .client
+            .get(url)
+            .send()
+            .await?
+            .json::<BlocklistStatus>()
+            .await?;
+
+        // Check if the request can be accepted or not based on the response
+        Ok(resp.accept)
+    }
+}
+
+impl BlocklistClient {
+    pub fn new() -> Self {
+        let base_url = format!(
+            "http://{}:{}",
+            SETTINGS.blocklist_client.host, SETTINGS.blocklist_client.port
+        );
+        BlocklistClient {
+            base_url,
+            client: Client::new(),
+        }
+    }
+    #[cfg(test)]
+    fn with_base_url(base_url: String) -> Self {
+        BlocklistClient {
+            base_url,
+            client: Client::new(),
+        }
+    }
+
+    fn address_screening_path(base_url: &str, address: &str) -> String {
+        format!("{}{}/{}", base_url, SCREEN_PATH, address)
+    }
+}
+
+impl Default for BlocklistClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mockito::{Server, ServerGuard};
+    use serde_json::json;
+    use tokio::sync::Mutex;
+
+    const ADDRESS: &str = "0x2337bBCD5766Bf2A9462D493E9A459b60b41B7f2";
+
+    struct TestContext {
+        server_guard: Mutex<ServerGuard>,
```

Comment:
What purpose does the `Mutex` serve here? It looks to me like we could omit it entirely.